ctapipe/ml/apply.py

```diff
@@ -0,0 +1,59 @@
+import numpy as np
+from astropy.table import Table
+from ctapipe.containers import ArrayEventContainer, ReconstructedEnergyContainer
+from ctapipe.core import Component
+
+from .sklearn import Regressor
+
+
+class EnergyReconstructor(Component):
+    """Base class for energy estimation."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def __call__(self, event: ArrayEventContainer) -> ReconstructedEnergyContainer:
+        return ReconstructedEnergyContainer()
+
+
+class EnergyRegressor(EnergyReconstructor):
+    """sklearn-based energy regression"""
+
+    # TODO: models per tel type
+    #       but this needs to be done in the Trainer first
+
+    def __init__(self, features, model_path, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.model = Regressor.load(model_path)
+        # TODO: update model config
+        self.model.model.n_jobs = -1
+        self.model.model.verbose = 0
+        # TODO: use config system
+        self.features = features
+
+    def __call__(self, event: ArrayEventContainer) -> ReconstructedEnergyContainer:
+        """EventSource Loop"""
+        for tel_id in event.trigger.tels_with_trigger:
+            features = dict()
+
+            for name, container in event.dl1.tel[tel_id].parameters.as_dict().items():
+                for key, value in container.items():
+                    # TODO: find a better way of creating these values (prefix?)
+                    features.update({f"{name}_{key}": [value]})
+            # TODO: get all possible features
+
+            feature_array = Table(features)
+
+            # TODO: since a single nan is enough for a single row to be discarded,
+            #       find it earlier for small performance reasons
+            if np.count_nonzero(feature_array.to_pandas().isna()):
+                continue
+            else:
+                prediction = self.model.predict(feature_array)
+                event.dl2.tel[tel_id].energy.energy = prediction
+                event.dl2.tel[tel_id].energy.is_valid = True
+                event.dl2.tel[tel_id].energy.tel_ids = [tel_id]
+
```

Comment:
Should we also do the stereo prediction here as well ?  In general it would just be a function to call that takes the mono predictions + image parameters and does some sort of weighted average to produce the final prediction.  However, the user might want to be able to choose from several methods, so *maybe a separate Component run afterwards is better*.  What do you think?  (maybe you were already considering that)

A few more things to think about (with no real idea of how to support them):
- There also may be some ML algorithms that input the mono features for all telescopes and output directly a stereo prediction without doing mono predictions first (e.g. a recurrent neural network could potentially do that).  
- Some reconstruction algorithms predict the energy without machine learning (I guess that is ok, they would just write to the appropriate place in the Event Structure).  We might consider if using those predictions as features could still be interesting: e.g. include the ImPACT energy prediction in a Hillas-style energy regressor?

Reply:
I would separate the actual mono prediction from combining them into a stereo prediction because there might be several methods to do this and even algorithms that might even directly give a stereo prediction (like impact)

Reply:
yes, that's basically what I said above - I think it is the way to go.  Probably a skeleton of this should be included in this PR then and we can see how it would work.  The initial implement could be super simple, just take the mono predictions and weight by `hillas.intensity`